### PR TITLE
[CF-490] Support for cinder migration with re-attach of backend to destination Cloud

### PR DESCRIPTION
**Pull request by SVilgelm**
[CF-496] Use juno nfs folders for all boxes
[CF-498] Migrate volume with reattaching old files
[CF-490] Make a dump of databases
[CF-490] Mute stdout of fabric
[CF-490] Use override rules to override volume_type
[CF-490] Delete failed volume

### CODE REVIEW

**Comment by MirantisWorkloadMobility** [comment]
There were no python files updated. Admins, please, check pull request.

File: cloudferry_devlab/provision/configure_nfs.sh [hunk 1]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env bash
+
+CINDER_CONF_FILES="/etc/cinder/nfsshares*"
+
+sed -i "s/${1}/juno/g" ${CINDER_CONF_FILES}
```
Comment (roman-verchikov) [inline cloudferry_devlab/provision/configure_nfs.sh]: `${1}` is unreadable, please assign to a local variable

File: cloudferry_devlab/Vagrantfile [hunk 2]
```diff
@@ -101,6 +101,12 @@ Vagrant.configure(2) do |config|
             thisnode.vm.provision "shell", path: "./provision/cleanup_nova_instances.sh"
           end
           thisnode.vm.provision "shell", path: "./provision/libvirt.sh"
+
+          if ENV["SAME_NFS_SHARE"] and nodename != "juno" then
```
Comment (roman-verchikov) [inline cloudferry_devlab/Vagrantfile]: I'd suggest renaming to `SHARED_NFS_BACKEND`

File: cloudferry_devlab/provision/configure_nfs.sh (hunk 1 above)
Comment (SVilgelm) [inline cloudferry_devlab/provision/configure_nfs.sh]: Done

File: cloudferry_devlab/Vagrantfile (hunk 2 above)
Comment (SVilgelm) [inline cloudferry_devlab/Vagrantfile]: Done

**Comment by MirantisWorkloadMobility** [comment]
There were no python files updated. Admins, please, check pull request.

**Comment by roman-verchikov** [comment]
+1

**Comment by MirantisWorkloadMobility** [comment]
There were no python files updated. Admins, please, check pull request.

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferry.lib.os.storage.cinder_db
cloudferry/lib/os/storage/cinder_db.py:23: [W0621(redefined-outer-name), 
CinderVolume.__init__] Redefining name 'uuid' from outer scope (line 15)
************* Module cloudferry.cfglib
cloudferry/cfglib.py:692: [E0602(undefined-variable), ] Undefined variable 'snapshot'
cloudferry/cfglib.py:692: [E0602(undefined-variable), ] Undefined variable 
'snapshot_opts'

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/3817/

File: cloudferry/actions/block_storage/migrate_volumes.py [hunk 3]
```diff
@@ -116,6 +122,34 @@ def __init__(self, init):
...
+    def _is_nfs_shared(self):
+        src_shared_nfs = isinstance(self.src_cinder_backend,
+                                    generic.SharedNFSPlugin)
+        dst_shared_nfs = isinstance(self.dst_cinder_backend,
+                                    generic.SharedNFSPlugin)
+        if src_shared_nfs and dst_shared_nfs:
+            return True
+        elif src_shared_nfs or dst_shared_nfs:
+            msg = ("Invalid configuration of src or dst storage backends. "
+                   "In case OF shared-nfs both options of the backends must "
```
Comment (roman-verchikov) [inline cloudferry/actions/block_storage/migrate_volumes.py]: Change "OF" to lowercase

File: cloudferry/actions/block_storage/migrate_volumes.py [hunk 4]
```diff
@@ -215,13 +249,46 @@ def run(self, **kwargs):
...
-            view.after_migration(i, volume)
-            new_volumes[volume['id']] = migrated_volume
+        for i, src_volume in enumerate(volumes):
+            view.before_migration(i, src_volume)
+
+            LOG.info("Checking if volume '%s' already present in destination",
+                     volume_name(src_volume))
+            dst_cinder = self.dst_cloud.resources[utils.STORAGE_RESOURCE]
+
+            dst_volume = dst_cinder.get_migrated_volume(src_volume['id'])
```
Comment (roman-verchikov) [inline cloudferry/actions/block_storage/migrate_volumes.py]: You'd need to modify `get_migrated_volume` to look for volume with the same ID for 'shared-nfs' case, and keep existing implementation for other backends.

File: cloudferry/cfglib.py [hunk 5]
```diff
@@ -309,7 +309,7 @@
 
 src_storage_opts = [
     cfg.StrOpt('backend', default='nfs',
-               choices=['nfs', 'iscsi-vmax'],
+               choices=['nfs', 'iscsi-vmax', 'shared-nfs'],
```
Comment (roman-verchikov) [inline cloudferry/cfglib.py]: Please update documentation too

File: cloudferry/lib/utils/mysql_connector.py [hunk 6]
```diff
@@ -36,6 +41,29 @@ def get_db_host(cloud_config):
...
+    db_host = get_db_host(cloud.cloud_config)
+    rr = remote_runner.RemoteRunner(
+        db_host, cloud.cloud_config.cloud.ssh_user,
+        password=cloud.cloud_config.cloud.ssh_password)
+    dump = rr.run(cmd,
+                  database=database,
+                  user=cloud.cloud_config.mysql.db_user,
+                  password=cloud.cloud_config.mysql.db_password)
+
+    filename = "dump_{position}_{database}_{time}.sql"
```
Comment (roman-verchikov) [inline cloudferry/lib/utils/mysql_connector.py]: I'd suggest this value to be configurable

File: cloudferry/actions/block_storage/migrate_volumes.py (hunk 3 above)
Comment (SVilgelm) [inline cloudferry/actions/block_storage/migrate_volumes.py]: Done

File: cloudferry/cfglib.py (hunk 5 above)
Comment (SVilgelm) [inline cloudferry/cfglib.py]: Done

File: cloudferry/actions/block_storage/migrate_volumes.py (hunk 4 above)
Comment (SVilgelm) [inline cloudferry/actions/block_storage/migrate_volumes.py]: No, it works fine. It uses metadata to match the  volumes.

File: cloudferry/lib/utils/mysql_connector.py (hunk 6 above)
Comment (SVilgelm) [inline cloudferry/lib/utils/mysql_connector.py]: Done

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/3821/

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by SVilgelm** [comment]
It works: http://jenkins-cf.vm.mirantis.net:8080/view/Functional%20test/job/cloudferry-functional-tests-grizzly-to-juno-shared-nfs/26/

**Comment by SVilgelm** [comment]
retest

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

File: cloudferry/lib/utils/mysql_connector.py [hunk 7]
```diff
@@ -36,6 +41,29 @@ def get_db_host(cloud_config):
     return db_host
 
 
+def dump_db(cloud, database=ALL_DATABASES):
+    cmd = "mysqldump {database} --user={user}"
```
Comment (roman-verchikov) [inline cloudferry/lib/utils/mysql_connector.py]: This command produces tons of unnecessary logs, please consider calling `mysqldump` remotely

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/3844/

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferry.lib.utils.log
cloudferry/lib/utils/log.py:45: [W0212(protected-access), StdoutLogger.write] Access to 
a protected member _mute of a client class
cloudferry/lib/utils/log.py:23: [W0611(unused-import), ] Unused import threading

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferry.lib.utils.log
cloudferry/lib/utils/log.py:44: [W0212(protected-access), StdoutLogger.write] Access to 
a protected member _mute of a client class

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by SVilgelm** [comment]
http://jenkins-cf.vm.mirantis.net:8080/view/Functional%20test/job/cloudferry-functional-tests-grizzly-to-juno-shared-nfs/34/

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

File: cloudferry/lib/utils/override.py [hunk 8]
```diff
@@ -70,18 +70,29 @@ def __init__(self, mapping):
...
+
         with open(mapping_filename, 'r') as f:
             data = yaml.load(f)
             if not isinstance(data, dict):
                 raise TypeError('%s root object must be dictionary!' %
                                 (mapping_filename,))
-        return cls(data)
+            data = data.get(object_type, {})
+            if not isinstance(data, dict):
+                raise TypeError('%s root object for %s must be dictionary!' %
```
Comment (antonf) [inline cloudferry/lib/utils/override.py]: this is not root object

File: cloudferry/lib/utils/override.py [hunk 9]
```diff
@@ -70,18 +70,29 @@ def __init__(self, mapping):
...
+        if mapping_filename is None:
+            return cls.zero()
+
         with open(mapping_filename, 'r') as f:
             data = yaml.load(f)
             if not isinstance(data, dict):
                 raise TypeError('%s root object must be dictionary!' %
                                 (mapping_filename,))
-        return cls(data)
+            data = data.get(object_type, {})
```
Comment (antonf) [inline cloudferry/lib/utils/override.py]: Please use different variable name

**Comment by antonf** [comment]
👎

File: cloudferry/lib/os/storage/cinder_storage.py [hunk 10]
```diff
@@ -341,6 +345,11 @@ def create_volume(self, size, **kwargs):
         cinder = self.cinder_client
         tenant_id = kwargs.get('project_id')
 
+        if self.cloud.position == 'dst':
```
Comment (roman-verchikov) [inline cloudferry/lib/os/storage/cinder_storage.py]: I think `volume_type` should be treated equally for both source and destination, I'd suggest removing this check

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

File: cloudferry/lib/utils/override.py (hunk 9 above)
Comment (SVilgelm) [inline cloudferry/lib/utils/override.py]: Done

File: cloudferry/lib/utils/override.py (hunk 8 above)
Comment (SVilgelm) [inline cloudferry/lib/utils/override.py]: Done

File: cloudferry/lib/os/storage/cinder_storage.py (hunk 10 above)
Comment (SVilgelm) [inline cloudferry/lib/os/storage/cinder_storage.py]: OK. Done

**Comment by MirantisWorkloadMobility** [comment]
$PULL_STATUS
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/3859/

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by SVilgelm** [comment]
retest

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
There were no python files updated. Admins, please, check pull request.

File: cloudferry/lib/utils/override.py [hunk 11]
```diff
@@ -70,18 +70,29 @@ def __init__(self, mapping):
...
+
         with open(mapping_filename, 'r') as f:
             data = yaml.load(f)
             if not isinstance(data, dict):
                 raise TypeError('%s root object must be dictionary!' %
                                 (mapping_filename,))
-        return cls(data)
+            object_data = data.get(object_type, {})
+            if not isinstance(object_data, dict):
+                raise TypeError('%s object must be dictionary!' %
```
Comment (antonf) [inline cloudferry/lib/utils/override.py]: Please specify the file: '%s object in mapping file %s must be dictionary!'

**Comment by antonf** [comment]
👎

File: cloudferry/lib/utils/override.py (hunk 11 above)
Comment (SVilgelm) [inline cloudferry/lib/utils/override.py]: Done

**Comment by MirantisWorkloadMobility** [comment]
$PYLINT_STATUS

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by roman-verchikov** [comment]
please rebase

**Comment by MirantisWorkloadMobility** [comment]
$PYLINT_STATUS

**Comment by MirantisWorkloadMobility** [comment]
$PULL_STATUS
$BUILD_URL

**Comment by roman-verchikov** [comment]
-1

1. We must not leave DB in undefined state in case when DB error occurs for any reason;
2. Rollback procedure for failed volume should be kept the same as previously: if failure occurs, created volume must be removed and warning should be printed.

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by SVilgelm** [comment]
works with shared-nfs: http://jenkins-cf.vm.mirantis.net:8080/view/Functional%20test/job/cloudferry-functional-tests-grizzly-to-juno-shared-nfs/39/

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by archyufa** [comment]
Migration of single unattached volume worked in staging.

**Comment by archyufa** [comment]
Can the backup of cinder databases be written to the file ? At the moment it is just dumps on screen.

**Comment by SVilgelm** [comment]
I'm working on it now

**Comment by archyufa** [comment]
Tested in staging. It looks like quota is not working correctly. As it only increase quota per 1 GB every time, even if volume is more than 1 GB.

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check failed
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/3941/

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check failed
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/3943/

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/3948/

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/3949/

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by SVilgelm** [comment]
retest

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by archyufa** [comment]
+1 tested in staging

**Comment by archyufa** [comment]
can be merged

File: configs/config.ini [hunk 12]
```diff
@@ -208,6 +208,12 @@ migrate_whole_cloud = False
 # Value is in seconds.
 image_save_timeout = 600
 
+# Filename template to store a dump of a database.
+db_dump_filename = dump_{position}_{database}_{time}.sql
+
+# Run mysqldump command on database's host. Otherwise mysqldump will be
```
Comment (archyufa) [inline configs/config.ini]: a bit not clear. If set to True it will run command on database's host?

Comment (SVilgelm) [inline configs/config.ini]: yes, or on the `migration.mysqldump_host` if it is set.

**Comment by archyufa** [comment]
It will be good to have possibility to disable snapshot. In case problem with it. Also there will be no need to take snapshots every time as this dumps consume space and takes extra time to run every time.

File: cloudferry/actions/block_storage/migrate_volumes.py [hunk 13]
```diff
@@ -205,24 +232,74 @@ def _create_volume(self, src_volume):
...
+                LOG.info("Volume '%s' is already present in destination "
+                         "cloud, skipping", src_volume['id'])
+            else:
+                try:
+                    dst_volume = self.migrate_func(src_volume)
+                except (plugins.base.VolumeObjectNotFoundError,
+                        retrying.TimeoutExceeded,
+                        exception.TenantNotPresentInDestination,
+                        cinder_exceptions.OverLimit,
+                        cinder_exceptions.NotFound,
```
Comment (roman-verchikov) [inline cloudferry/actions/block_storage/migrate_volumes.py]: Let's catch all cinder exceptions here

**Comment by roman-verchikov** [comment]
-1, please update exception handling

**Comment by antonf** [comment]
> It will be good to have possibility to disable snapshot

Are you serious? I am surprised that person working with production environment request such a "feature". In case there is problem with making database snapshot you should fix this instead of disabling making backup of your database. Working with OpenStack database directly is extremely dangerous and you should have backup of production database as recent as possible *each time*. Production database can change at any time: even when you run your migration for tenant A, some user can create volume in tenant B, for example.

While it is possible to do what you ask, we shouldn't do that.

File: cloudferry/actions/block_storage/migrate_volumes.py (hunk 13 above)
Comment (SVilgelm) [inline cloudferry/actions/block_storage/migrate_volumes.py]: Done

**Comment by SVilgelm** [comment]
retest

**Comment by SVilgelm** [comment]
It works: http://jenkins-cf.vm.mirantis.net:8080/view/Functional%20test/job/cloudferry-functional-tests-grizzly-to-juno-shared-nfs/53/

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by roman-verchikov** [comment]
can be merged

**Pull request merged**